File: pa3/src/List.java
import java.util.Objects;

/*********************************************************************
 * Malcolm Riley
 *
 * CruzID: masriley
 * Assignment: pa3
 *
 * 10-2017
 *********************************************************************/

public class List<K> {

	private static final int CURSOR_INDEX_INVALID = -1;

	private Node<K> NODE_CURSOR;
	private Node<K> NODE_FRONT;
	private Node<K> NODE_BACK;
	private int INDEX = CURSOR_INDEX_INVALID;
	private int LENGTH;

	public List() {

	}

	/* Required Methods */

	public int length() {
		return this.LENGTH;
	}

	public int index() {
		if (this.isNodeDefined(this.NODE_CURSOR)) {
			return this.INDEX;
		}
		return -1;
	}

	public K front() {
		return this.NODE_FRONT.get();
	}

	public K back() {
		return this.NODE_BACK.get();
	}

	public K get() {
		return this.NODE_CURSOR.get();
	}
	
	/**
	 * Overrides {@link Object#equals(Object)}.
	 */
	public boolean equals(Object passedList) {
		if (passedList instanceof List<?>) {
			return Objects.equals(this, passedList);
		}
		return false;
	}

	/**
	 * Determines whether this {@link List} is equivalent to the passed {@link List}; that is, whether it contains equivalent values in the same order.
	 * 
	 * @param passedList - The list to compare this one to
	 * @return Whether or not the two lists are equivalent.
	 */
	public boolean equals(List<?> passedList) {
		if (this.length() == passedList.length()) {
			if (!this.isEmpty()) {
				List<K>.Node<K> thisListIterator = this.getFrontNode();
				List<?>.Node<?> passedListIterator = passedList.getFrontNode();
				while (isNodeDefined(thisListIterator)) {
					if (!Objects.equals(thisListIterator, passedListIterator)) {
						return false;
					}
					thisListIterator = thisListIterator.getNext();
					passedListIterator = passedListIterator.getNext();
				}
			}
			return true;
		}
		return false;
	}

	/**
	 * Clears the list, removing and resetting all owned {@link Node}s.
	 */
	public void clear() {
		while (!this.isEmpty()) {
			this.deleteBack();
		}
		this.NODE_FRONT = null;
		this.NODE_BACK = null;
		this.NODE_CURSOR = null;
		this.INDEX = CURSOR_INDEX_INVALID;
		this.LENGTH = 0;
	}

	/**
	 * Moves the cursor to the front of the {@link List}.
	 */
	public void moveFront() {
		if (this.isNodeDefined(this.NODE_FRONT)) {
			this.NODE_CURSOR = this.NODE_FRONT;
			this.INDEX = 0;
		}
	}

	/**
	 * Moves the cursor to the back of the {@link List}.
	 */
	public void moveBack() {
		if (this.isNodeDefined(this.NODE_BACK)) {
			this.NODE_CURSOR = this.NODE_BACK;
			this.INDEX = (this.LENGTH - 1);
		}
	}

	/**
	 * Moves the cursor to the position one previous from its current position in this {@link List}.
	 */
	public void movePrev() {
		if (this.setAndCheckCursor(this.NODE_CURSOR.getPrevious())) {
			this.INDEX -= 1;
			if (this.INDEX < 0) {
				this.INDEX = CURSOR_INDEX_INVALID;
			}
		}
	}

	/**
	 * Moves the cursor to the position one next from its current position in this {@link List}.
	 */
	public void moveNext() {
		if (this.setAndCheckCursor(this.NODE_CURSOR.getNext())) {
			this.INDEX += 1;
			if (this.INDEX > (this.LENGTH - 1)) {
				this.INDEX = CURSOR_INDEX_INVALID;
			}
		}
	}

	public void prepend(K passedData) {
		Node<K> newNode = new Node<K>(this, passedData);
		if(this.insertNodeBefore(this.NODE_FRONT, newNode)) {
			this.INDEX += 1;
		}
		this.NODE_FRONT = newNode;
	}

	public void append(K passedData) {
		Node<K> newNode = new Node<K>(this, passedData);
		this.insertNodeAfter(this.NODE_BACK, newNode);
		this.NODE_BACK = newNode;
	}

	public void insertBefore(K passedData) {
		Node<K> newNode = new Node<K>(this, passedData);
		if (this.insertNodeBefore(this.NODE_CURSOR, newNode)) {
			this.INDEX += 1;
		}
	}

	public void insertAfter(K passedData) {
		Node<K> newNode = new Node<K>(this, passedData);
		this.insertNodeAfter(this.NODE_CURSOR, newNode);
	}

	public void deleteFront() {
		if (this.NODE_CURSOR == this.NODE_FRONT) {
			this.INDEX = CURSOR_INDEX_INVALID;
		}
		else {
			this.INDEX -= 1;
		}
		this.onNodeRemoved(this.NODE_FRONT);
	}

	public void deleteBack() {
		if (this.NODE_CURSOR == this.NODE_BACK) {
			this.INDEX = CURSOR_INDEX_INVALID;
		}
		this.onNodeRemoved(this.NODE_BACK);
	}

	public void delete() {
		this.onNodeRemoved(this.NODE_CURSOR);
	}

	public List<K> copy() {
		List<K> newList = new List<K>();
		Node<K> thisListIterator = this.getFrontNode();
		while (this.isNodeDefined(thisListIterator)) {
			newList.append(thisListIterator.get());
			thisListIterator = thisListIterator.getNext();
		}
		return newList;
	}

	public List<K> concat(List<K> passedList) {
		List<K> newList = new List<K>();
		Node<K> thisListIterator = this.getFrontNode();
		while (this.isNodeDefined(thisListIterator)) {
			newList.append(thisListIterator.get());
			thisListIterator = thisListIterator.getNext();
		}
		return null;
	}

	@Override
	public String toString() {
		String stringRepresentation = "";
		Node<K> thisListIterator = this.getFrontNode();
		while (this.isNodeDefined(thisListIterator)) {
			stringRepresentation += thisListIterator.toString();
			thisListIterator = thisListIterator.getNext();
			if (thisListIterator != null) {
				stringRepresentation += " ";
			}
		}
		return stringRepresentation;
	}

	protected Node<K> getFrontNode() {
		return this.NODE_FRONT;
	}

	protected Node<K> getBackNode() {
		return this.NODE_BACK;
	}

	protected Node<K> getCursorNode() {
		return this.NODE_CURSOR;
	}

	protected boolean isEmpty() {
		return (this.LENGTH < 1);
	}

	protected boolean isNodeDefined(Node<?> passedNode) {
		return (passedNode != null) && (passedNode.get() != null) && (passedNode.getOwner() == this);
	}
	
	private boolean insertNodeBefore(Node<K> passedExistingNode, Node<K> passedInsertedNode) {
		this.onNodeAdded(passedInsertedNode);
		if (isNodeDefined(passedExistingNode)) {
			// Set existing links
			Node<K> previous = passedExistingNode.getPrevious();
			if (isNodeDefined(previous)) {
				previous.setNext(passedInsertedNode);
			}
			passedExistingNode.setPrevious(passedInsertedNode);
			
			// Set new links
			passedInsertedNode.setNext(passedExistingNode);
			passedInsertedNode.setPrevious(previous);

			// Set new front if applicable
			if (this.NODE_FRONT == passedExistingNode) {
				this.NODE_FRONT = passedInsertedNode;
			}
			return true;
		}
		return false;
	}

	private boolean insertNodeAfter(Node<K> passedExistingNode, Node<K> passedInsertedNode) {
		this.onNodeAdded(passedInsertedNode);
		if (isNodeDefined(passedExistingNode)) {
			// Set existing links
			Node<K> next = passedExistingNode.getNext();
			if (isNodeDefined(next)) {
				next.setPrevious(passedInsertedNode);
			}
			passedExistingNode.setNext(passedInsertedNode);
			// Set new links
			passedInsertedNode.setPrevious(passedExistingNode);
			passedInsertedNode.setNext(next);
			
			// Set new back if applicable
			if (this.NODE_BACK == passedExistingNode) {
				this.NODE_BACK = passedInsertedNode;
			}
			return true;
		}
		return false;
	}
	
	private void onNodeAdded(Node<K> passedNode) {
		if (this.isEmpty()) {
			this.NODE_FRONT = passedNode;
			this.NODE_BACK = passedNode;
		}
		this.LENGTH += 1;
	}

	private void onNodeRemoved(Node<K> passedNode) {
		if (this.isNodeDefined(passedNode)) {
			Node<K> nextNode = passedNode.getNext();
			Node<K> previousNode = passedNode.getPrevious();
			if (this.isNodeDefined(nextNode)) {
				nextNode.setPrevious(passedNode.getPrevious());
				if (passedNode == this.NODE_FRONT) {
					this.NODE_FRONT = nextNode;
				}
			}
			if (this.isNodeDefined(previousNode)) {
				previousNode.setNext(passedNode.getNext());
				if (passedNode == this.NODE_BACK) {
					this.NODE_BACK = previousNode;
				}
			}
			passedNode.reset();
			passedNode = null;
			this.LENGTH -= 1;
		}
	}
	
	private boolean setAndCheckCursor(Node<K> passedNode) {
		this.NODE_CURSOR = passedNode;
		if (!isNodeDefined(this.NODE_CURSOR)) {
			this.INDEX = CURSOR_INDEX_INVALID;
			return false;
		}
		return true;
	}

	/* Node Implementation */

	private class Node<T> {
		private final List<T> owningList;
		private T element;
		private Node<T> previous;
		private Node<T> next;

		public Node(List<T> passedOwningList, T passedElement) {
			this.owningList = passedOwningList;
			this.element = passedElement;
		}

		public T get() {
			return this.element;
		}

		protected List<T> getOwner() {
			return this.owningList;
		}

		protected void reset() {
			this.element = null;
			this.previous = null;
			this.next = null;
		}

		protected Node<T> getNext() {
			return this.next;
		}

		protected Node<T> getPrevious() {
			return this.previous;
		}

		protected void setNext(Node<T> passedNode) {
			this.next = passedNode;
		}

		protected void setPrevious(Node<T> passedNode) {
			this.previous = passedNode;
		}
		
		@Override
		public boolean equals(Object passedObject) {
			if (passedObject instanceof List<?>.Node<?>) {
				Object thisValue = this.get();
				Object passedValue = ((List<?>.Node<?>) passedObject).get();
				return Objects.equals(thisValue, passedValue);
			}
			return false;
		}

		@Override
		public String toString() {
			return this.element.toString();
		}
	}

}

File: pa4/src/List.c
/*********************************************************************
 * Malcolm Riley
 *
 * CruzID: masriley
 * Assignment: pa4, modified from pa2
 *
 * 11-2017
 *********************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include "List.h"

/* Internal Function Declarations */
void exitBadWithMessage(const char* passedCharArray);
void insertIntoEmpty(List passedList, Node passedNode);
int insertNodeBefore(List passedList, Node passedNode, Node passedInsertedNode);
int insertNodeAfter(List passedList, Node passedNode, Node passedInsertedNode);
int removeNode(List passedList, Node passedNode);
int isNull(void* passedPointer, char* passedCharArray, int passedIsTerminal);
int isListEmpty(List passedList, char* passedCharArray, int passedIsTerminal);
int validateIndex(List passedList);
int isCursorValid(List passedList, char* passedCharArray, int passedIsTerminal);

Node newNode(int passedValue) {
	Node newNode = malloc(sizeof(NodeObject));
	newNode->nextNode = NULL;
	newNode->previousNode = NULL;
	newNode->value = passedValue;
	return newNode;
}

void freeNode(Node* passedNode) {
	(*passedNode)->nextNode = NULL;
	(*passedNode)->previousNode = NULL;
	// free(passedNode);
	free(*passedNode);
}

/* Node Functions */

int nodesAreEqual(Node passedFirstNode, Node passedSecondNode) {
	if ((passedFirstNode != NULL) && (passedSecondNode != NULL)) {
		return (passedFirstNode->value == passedSecondNode->value);
	}
	return FALSE;
}

/* List Constructor/Destructor */
List newList(void) {
	List newList = malloc(sizeof(ListObject));
	newList->cursorIndex = UNDEFINED;
	newList->length = 0;
	newList->nodeBack = NULL;
	newList->nodeFront = NULL;
	newList->nodeCursor = NULL;
	return newList;
}

void freeList(List* passedList) {
	clear(*passedList);
	(*passedList)->nodeBack = NULL;
	(*passedList)->nodeFront = NULL;
	(*passedList)->nodeCursor = NULL;
	// free(passedList);
	free(*passedList);
}

/* List Functions */
int length(List passedList) {
	if (!isNull(passedList, "Cannot retrieve the length of a null List.", TRUE)) {
		return passedList->length;
	}
	return UNDEFINED;
}

int index(List passedList) {
	if (!isNull(passedList, "Cannot retrieve the index of a null List.", TRUE)) {
		validateIndex(passedList);
		return passedList->cursorIndex;
	}
	return UNDEFINED;
}

int front(List passedList) {
	if (!isNull(passedList, "Cannot retrieve the front of a null List.", TRUE)) {
		if (!isNull(passedList->nodeFront, "Front Node is NULL.", FALSE)) {
			return passedList->nodeFront->value;
		}
	}
	return UNDEFINED;
}


int back(List passedList) {
	if (!isNull(passedList, "Cannot retrieve the back of a null List.", TRUE)) {
		if (!isNull(passedList->nodeBack, "Back Node is NULL.", FALSE)) {
			return passedList->nodeBack->value;
		}
	}
	return UNDEFINED;
}

int get(List passedList) {
	if (!isNull(passedList, "Cannot retrieve the cursor of a null List.", TRUE)) {
		if (!isNull(passedList->nodeCursor, "Cursor Node is NULL.", FALSE)) {
			return passedList->nodeCursor->value;
		}
	}
	return UNDEFINED;
}

int equals(List passedFirstList, List passedSecondList) {
	if ((!isNull(passedFirstList, "Equals: First list is null.", FALSE)) && (!isNull(passedSecondList, "Equals: Second list is null.", FALSE))) {
		int areEqual = (passedFirstList->length == passedSecondList->length);
		if (areEqual) {
			Node firstNode = passedFirstList->nodeFront;
			Node secondNode = passedSecondList->nodeFront;
			for(int ii = 0; (ii < passedFirstList->length) && (areEqual); ii += 1) {
				areEqual = nodesAreEqual(firstNode, secondNode);
				firstNode = firstNode->nextNode;
				secondNode = secondNode->nextNode;
			}
		}
		return areEqual;
	}
	return FALSE;
}

void clear(List passedList) {
	while (passedList->length > 0) {
		deleteBack(passedList);
	}
}

void moveFront(List passedList) {
	if (!isListEmpty(passedList, "Error with List when moveFront was called:", TRUE)) {
		if (!isNull(passedList->nodeFront, "Error with List's front Node:", FALSE)) {
			passedList->nodeCursor = passedList->nodeFront;
			passedList->cursorIndex = 0;
		}
	}
}

void moveBack(List passedList) {
	if (!isListEmpty(passedList, "Error with List when moveBack was called:", TRUE)) {
		if (!isNull(passedList->nodeBack, "Error with List's back Node:", FALSE)) {
			passedList->nodeCursor = passedList->nodeBack;
			passedList->cursorIndex = (length(passedList) - 1);
		}
	}
}

void movePrev(List passedList) {
	if (!isListEmpty(passedList, "Error with List when movePrev was called:", FALSE)) {
		if (!isNull(passedList->nodeCursor, "Error with List's Cursor Node:", FALSE)) {
			passedList->nodeCursor = passedList->nodeCursor->previousNode;
			passedList->cursorIndex -= 1;
		}
	}
}

void moveNext(List passedList) {
	if (!isListEmpty(passedList, "Error with List when moveNext was called:", FALSE)) {
		if (!isNull(passedList->nodeCursor, "Error with List's Cursor Node:", FALSE)) {
			passedList->nodeCursor = passedList->nodeCursor->nextNode;
			passedList->cursorIndex += 1;
		}
	}
}

void prepend(List passedList, int passedValue) {
	if (!isNull(passedList, "Cannot prepend to a null List!", TRUE)) {
		Node allocatedNode = newNode(passedValue);
		if (passedList->length < 1) {
			insertIntoEmpty(passedList, allocatedNode);
		}
		else if (insertNodeBefore(passedList, passedList->nodeFront, allocatedNode)) {
			passedList->cursorIndex += 1;
			passedList->length += 1;
			passedList->nodeFront = allocatedNode;
		}
	}
}

void append(List passedList, int passedValue) {
	if (!isNull(passedList, "Cannot append to a null List!", TRUE)) {
		Node allocatedNode = newNode(passedValue);
		if (passedList->length < 1) {
			insertIntoEmpty(passedList, allocatedNode);
		}
		else if (insertNodeAfter(passedList, passedList->nodeBack, allocatedNode)) {
			passedList->length += 1;
			passedList->nodeBack = allocatedNode;
		}
	}
}

void insertBefore(List passedList, int passedValue) {
	if (!isNull(passedList, "Cannot insert into a null List!", TRUE)) {
		if (isCursorValid(passedList, "Cannot insert before an undefined cursor!", TRUE)) {
			Node allocatedNode = newNode(passedValue);
			if (insertNodeBefore(passedList, passedList->nodeCursor, allocatedNode)) {
				passedList->cursorIndex += 1;
				passedList->length += 1;
			}
		}
	}
}

void insertAfter(List passedList, int passedValue) {
	if (!isNull(passedList, "Cannot insert into a null List!", TRUE)) {
		if (isCursorValid(passedList, "Cannot insert after an undefined cursor!", TRUE)) {
			Node allocatedNode = newNode(passedValue);
			if (insertNodeAfter(passedList, passedList->nodeCursor, allocatedNode)) {
				passedList->length += 1;
			}
		}
	}
}

void deleteFront(List passedList) {
	if (!isListEmpty(passedList, "Error when deleting the front Node of a List", TRUE)) {
		if (removeNode(passedList, passedList->nodeFront)) {
			passedList->cursorIndex -= 1;
			passedList->length -= 1;
		}
	}
}

void deleteBack(List passedList) {
	if (!isListEmpty(passedList, "Error when deleting the back Node of a List", TRUE)) {
		if (removeNode(passedList, passedList->nodeBack)) {
			passedList->length -= 1;
		}
	}
}

void delete(List passedList) {
	if (!isListEmpty(passedList, "Error when deleting the cursor Node of a List", TRUE)) {
		if (removeNode(passedList, passedList->nodeCursor)) {
			passedList->length -= 1;
			passedList->cursorIndex = UNDEFINED;
		}
	}
}

List concatList(List passedFirstList, List passedSecondList) {
	int firstListValid = !isListEmpty(passedFirstList, "Error with prepended List during concatenation:", FALSE);
	int secondListValid = !isListEmpty(passedFirstList, "Error with appended List during concatenation:", FALSE);
	List allocatedList = newList();
	// If both Lists are valid, concatenate
	if ((firstListValid != FALSE) && (secondListValid != FALSE)) {
		List allocatedList = copyList(passedFirstList);
		Node iteratedNode = passedSecondList->nodeFront;
		while (iteratedNode != NULL) {
			append(allocatedList, iteratedNode->value);
			iteratedNode = iteratedNode->nextNode;
		}
		return allocatedList;
	}
	// If only first List is valid, use that one
	else if (firstListValid != FALSE) {
		return copyList(passedFirstList);
	}

	// If only second List is valid, use that one
	else if (secondListValid != FALSE) {
		return copyList(passedSecondList);
	}

	// Otherwise, return empty List
	return allocatedList;
}

void printList(FILE* passedOutputFile, List passedList) {
	if(!isNull(passedOutputFile, "Cannot print to a null file reference.", FALSE)) {
		if (!isListEmpty(passedList, "Error with List during print:", FALSE)) {
			Node iteratedNode = passedList->nodeFront;
			while (iteratedNode != NULL) {
				fprintf(passedOutputFile, "%d", iteratedNode->value);
				iteratedNode = iteratedNode->nextNode;
				if (iteratedNode != NULL) {
					fprintf(passedOutputFile, "%s", " ");
				}
			}
		}
	}
}

List copyList(List passedList) {
	List allocatedList = newList();
	if (!isListEmpty(passedList, "Cannot copy a null List.", FALSE)) {
		Node iteratedNode = passedList->nodeFront;
		while (iteratedNode != NULL) {
			append(allocatedList, iteratedNode->value);
			iteratedNode = iteratedNode->nextNode;
		}
	}
	return allocatedList;
}

/* New for pa4 */

/**
 * Inserts the passed value in sorted order into the passed List.
 *
 * This does not SORT the List, merely inserts into sorted order, so if the
 * List is not already sorted, this only has the effect of inserting the value
 * at the first place that it is less than the next value.
 */
void insertInto(List passedList, int passedValue) {
	if (!isNull(passedList, "Cannot insert into a null List!", TRUE)) {
		if (passedList->length == 0) {
			append(passedList, passedValue);
		}
		else {
			int value = -1;
			for (moveFront(passedList); get(passedList) > 0; moveNext(passedList)) {
				value = get(passedList);
				if (passedValue < value) {
					insertBefore(passedList, passedValue);
					return;
				}
			}
		}
	}
}

/* Internal Functions */

/**
 * If the passedList is empty, inserts into it, setting the front and back nodes.
 */
void insertIntoEmpty(List passedList, Node passedNode) {
	passedList->nodeFront = passedNode;
	passedList->nodeBack = passedNode;
	passedList->length += 1;
}

/**
 * Removes a Node from passedList, freeing it.
 */
int removeNode(List passedList, Node passedNode) {
	if (!isNull(passedList, "Cannot remove a Node from a List that is null.", FALSE)) {
		if (!isNull(passedNode, "Cannot remove a Node that is null.", FALSE)) {
			Node nextNode = passedNode->nextNode;
			Node previousNode = passedNode->previousNode;
			if (nextNode != NULL) {
				nextNode->previousNode = passedNode->previousNode;
				if (passedNode == passedList->nodeFront) {
					passedList->nodeFront = nextNode;
				}
			}
			if (previousNode != NULL) {
				previousNode->nextNode = passedNode->nextNode;
				if (passedNode == passedList->nodeBack) {
					passedList->nodeBack = previousNode;
				}
			}
			freeNode(&passedNode);
			return TRUE;
		}
	}
	return FALSE;
}

/**
 * Inserts passedInsertedNode before passedNode.
 */
int insertNodeBefore(List passedList, Node passedNode, Node passedInsertedNode) {
	int passedNodeValid = !isNull(passedNode, "Attempting to insert before null Node", FALSE);
	int insertedNodeValid = !isNull(passedNode, "Attempting to insert null Node", FALSE);
	if ((passedNodeValid) && (insertedNodeValid)) {
		// Set existing links
		Node previous = passedNode->previousNode;
		if (previous != NULL) {
			previous->nextNode = passedInsertedNode;
		}
		passedNode->previousNode = passedInsertedNode;

		// Set new links
		passedInsertedNode->nextNode = passedNode;
		passedInsertedNode->previousNode = previous;

		// Set new front if applicable
		if (passedList->nodeFront == passedNode) {
			passedList->nodeFront = passedInsertedNode;
		}
		return TRUE;
	}
	return FALSE;
}

/**
 * Inserts passedInsertedNode after the passedNode.
 */
int insertNodeAfter(List passedList, Node passedNode, Node passedInsertedNode) {
	int passedNodeValid = !isNull(passedNode, "Attempting to insert after null Node", FALSE);
	int insertedNodeValid = !isNull(passedNode, "Attempting to insert null Node", FALSE);
	if ((passedNodeValid) && (insertedNodeValid)) {
		// Set existing links
		Node next = passedNode->nextNode;
		if (next != NULL) {
			next->previousNode = passedInsertedNode;
		}
		passedNode->nextNode = passedInsertedNode;

		// Set new links
		passedInsertedNode->previousNode = passedNode;
		passedInsertedNode->nextNode = next;

		// Set new back if applicable
		if (passedList->nodeBack == passedNode) {
			passedList->nodeBack = passedInsertedNode;
		}
		return TRUE;
	}
	return FALSE;
}

/**
 * Checks whether the passed pointer is null, and prints the passed message in that case.
 *
 * If passedIsTerminal is specifically 1, calls exitBadWithMessage with an error message.
 */
int inline isNull(void* passedPointer, char* passedCharArray, int passedIsTerminal) {
	if (passedPointer == NULL) {
		puts(passedCharArray);
		if (passedIsTerminal == TRUE) {
			exitBadWithMessage("Fatal Error: Null pointer received.");
		}
		return TRUE;
	}
	return FALSE;
}

/**
 * Checks whether the passed List is empty. Does not perform null check on the passed List.
 */
int inline isListEmpty(List passedList, char* passedCharArray, int passedIsTerminal) {
	if (length(passedList) <= 0) {
		puts(passedCharArray);
		if (passedIsTerminal == TRUE) {
			exitBadWithMessage("Fatal Error: List is empty.");
		}
		return TRUE;
	}
	return FALSE;
}

/**
 * Validates whether the passedList's cursor is within a valid range (c > 0) && (c < length)
 */
int inline validateIndex(List passedList) {
	if ((passedList->cursorIndex < 0) || (passedList->cursorIndex >= passedList->length)) {
		passedList->cursorIndex = UNDEFINED;
	}
	return (passedList->cursorIndex >= 0);
}

/**
 * Checks whether the passed List's cursor is valid. Does not perform null check on the passed List.
 */
int inline isCursorValid(List passedList, char* passedCharArray, int passedIsTerminal) {
	if (!validateIndex(passedList)) {
		puts(passedCharArray);
		if (passedIsTerminal == TRUE) {
			exitBadWithMessage("Fatal Error: Cursor index is invalid.");
		}
		return FALSE;
	}
	if(passedList->nodeCursor == NULL) {
		puts(passedCharArray);
		if (passedIsTerminal == TRUE) {
			exitBadWithMessage("Fatal Error: Cursor is null.");
		}
		return FALSE;
	}
	return TRUE;
}

/**
 * Exits with a code of 1 after printing the passed message.
 */
void inline exitBadWithMessage(const char* passedCharArray) {
	puts(passedCharArray);
	exit(1);
}
